pkg/pdp/scheduler/handler.go

```diff
@@ -0,0 +1,265 @@
+
+	"github.com/storacha/storage/pkg/database"
+	"github.com/storacha/storage/pkg/pdp/service/models"
+)
+
+// taskTypeHandler ties a task implementation with engine-specific metadata.
+type taskTypeHandler struct {
+	TaskInterface
+	TaskTypeDetails TaskTypeDetails
+	TaskEngine      *TaskEngine
+	doneMu          sync.Mutex
+}
+
+// AddTask is the implementation passed to each task's Adder.
+// It creates a new task record in the database.
+func (h *taskTypeHandler) AddTask(extra func(TaskID, *gorm.DB) (bool, error)) {
+	var tID TaskID
+	retryWait := 100 * time.Millisecond
+
+retryAddTask:
+	err := h.TaskEngine.db.WithContext(h.TaskEngine.ctx).Transaction(func(tx *gorm.DB) error {
+		// Create a new Task record.
+		task := models.Task{
+			PostedTime: time.Now(),
+			AddedBy:    h.TaskEngine.sessionID,
+			Name:       h.TaskTypeDetails.Name,
+		}
+
+		// Insert the task and let GORM fill in the auto-generated ID.
+		if err := tx.Create(&task).Error; err != nil {
+			return fmt.Errorf("could not insert task: %w", err)
+		}
+
+		// Set the task ID from the newly inserted record.
+		tID = TaskID(task.ID)
+
+		// Call the extra callback to update additional info in the same transaction.
+		shouldCommit, err := extra(tID, tx)
+		if err != nil {
+			return err
+		}
+
+		if shouldCommit {
+			return nil
+		}
+
+		return ErrDoNotCommit
+	})
+	if err != nil {
+		// If a unique constraint error is detected, assume the task already exists.
+		if database.IsUniqueConstraintError(err) {
+			log.Debugf("addtask(%s) saw unique constraint, so it's added already.", h.TaskTypeDetails.Name)
+			return
+		}
+		// If it's a timeout error, backoff and retry.
+		if database.IsLockedError(err) {
+			log.Warnf("addtask(%s) saw locked error retrying in %s.", retryWait, h.TaskTypeDetails.Name)
+			time.Sleep(retryWait)
+			retryWait *= 2
+			goto retryAddTask
+		}
+
+		if errors.Is(err, ErrDoNotCommit) {
+			return
+		}
+		log.Errorw("Could not add task. AddTask func failed", "error", err, "type", h.TaskTypeDetails.Name)
+		return
+	}
+
+}
+
+// considerWork claims and executes tasks.
+func (h *taskTypeHandler) considerWork(taskIDs []TaskID, db *gorm.DB) bool {
+	acceptedAny := false
+
+	log.Debugf("Considering work for tasks %v", taskIDs)
+	for _, id := range taskIDs {
+		log.Debugf("Considering work for task %d", id)
+		result := db.Model(&models.Task{}).
+			Where(&models.Task{ID: int64(id), SessionID: nil}).
+			Updates(models.Task{
+				SessionID:  &h.TaskEngine.sessionID,
+				UpdateTime: time.Now(),
+			})
+
+		if result.Error != nil {
+			log.Errorw("Could not claim task", "task_id", id, "error", result.Error)
+			continue
+		}
+		if result.RowsAffected == 0 {
+			// Already taken by someone else (or in race condition). Skip it.
+			log.Debugf("Task %d was already claimed; skipping", id)
+			continue
+		}
+
+		// Successfully claimed this task, so let’s run it in a goroutine:
+		// TODO doing this in parallel is causing concurrency issues with the sqlite database.
```

Comment:
YOLO

Reply:
Actually, good call-out. This has been addressed by:
- the retry logic in `handleDoneTask` and `AddTask`, which will retry db operations that timeout due to locking until success.
- the `doneMu` of `taskTypeHandler` which locks around the db transaction.
- the changes to the `database` package which add `_pragma=journal_mode=wal` and `_pragma=busy_timeout=5000`, using a write ahead log and longer operation timeout. 

SQLite will still get cranky about parallel writes since it uses database-wide locks. 
Of note - previously this engine was designed to use Postgres (yugabyte) - which uses row-level locking. I think with the changes I've mentioned above, we should be good to remove the TODO here (which I'll do before merging)